src/table/concat_iterator.rs

```diff
@@ -0,0 +1,200 @@
+use super::iterator::ITERATOR_REVERSED;
+use super::{AgateIterator, Table, TableIterator};
+use crate::util::{KeyComparator, COMPARATOR};
+use crate::value::Value;
+
+use bytes::Bytes;
+
+/// ConcatIterator iterates on SSTs with no overlap keys.
+pub struct ConcatIterator {
+    cur: Option<usize>,
+    iters: Vec<Option<TableIterator>>,
+    tables: Vec<Table>,
+    opt: usize,
+}
+
+impl ConcatIterator {
+    /// Create `ConcatIterator` from a list of tables. Tables must have been sorted
+    /// and have no overlap keys.
+    pub fn from_tables(tables: Vec<Table>, opt: usize) -> Self {
+        let iters = tables.iter().map(|_| None).collect();
+
+        ConcatIterator {
+            cur: None,
+            iters,
+            tables,
+            opt,
+        }
+    }
+
+    fn set_idx(&mut self, idx: usize) {
+        if idx >= self.iters.len() {
+            self.cur = None;
+            return;
+        }
+        if self.iters[idx].is_none() {
+            self.iters[idx] = Some(self.tables[idx].new_iterator(self.opt));
+        }
+        self.cur = Some(idx);
+    }
+
+    fn iter_mut(&mut self, idx: usize) -> &mut TableIterator {
```

Comment:
seems the `idx` is always self.cur.unwrap(), can we remove the parameter?

Reply:
`self.cur` may become `None` when the iterator points to invalid location. So currently we could not remove this.

The underlying issue is that, badger use `-1` to represent iterator beyond reverse iteration, and we always use `usize` for iterator index. Therefore, we have to represent `cur` either in `Some/None` way, or represent `-1` as `usize::max` like what we've done in `TableIterator`.

Reply:
but now, `iter_mut` is always called when `self.cur` is some.

Reply:
Got it. I misunderstood previous comments. I'll remove the parameter from function signature.

Reply:
done